apps/src/code-studio/pd/components/useRegionalPartner.jsx

```diff
@@ -57,30 +57,29 @@ export const useRegionalPartner = data => {
   const [partner, setPartner] = useState(null);
   // cache this debounced function so that it is more easily testable
   const debouncedFetch = useCallback(
-    debounce(fetchRegionalPartner, 500, {leading: true}),
+    debounce(
+      data =>
+        fetchRegionalPartner(data)
+          .then(partner => {
+            // Update state with all the partner workshop data to display
+            setLoadingPartner(false);
+            setLoadError(false);
+            // the api returns an object with all fields set to null if not found
+            setPartner(partner.id === null ? null : partner); // TODO(tim): potential issue where requests are fulfilled out of order
+          })
+          .catch(() => {
+            setLoadingPartner(false);
+            setLoadError(true);
+          }),
+      500,
+      {leading: true}
+    ),
     []
```

Comment:
do we need to pass in `[data]` to make new requests get sent as the user keeps typing?  https://reactjs.org/docs/hooks-reference.html#usecallback

> every value referenced inside the callback should also appear in the dependencies array

Reply:
there's a special case usage of the dependency array, which is that when the array is empty, the function only gets executed on initialization. for useCallback, this lets us keep the same function (same reference) on each render. Otherwise (i.e. if we just defined the function like normal javascript), `debounce` would be called again on each render and would start a separate timeline of debounced calls and never actually do any debouncing.

Note that the useEffect call below does have all the variables that get used from data in its dependency array so that the debounced function is called each time something changes.

Reply:
(see the big yellow box at the bottom of the useEffect react docs for the special case [] array)

Reply:
💡 ! got it, you are debouncing the creation of the function, not the actual function call. that is what I was missing.

Reply:
sorry, you would be `memoizing` the creation of the function, not the actual function call. (you are still debouncing the calling of the function)